fvm/evm/offchain/blocks/block_context.go

```diff
@@ -0,0 +1,107 @@
+package blocks
+
+import (
+	gethCommon "github.com/onflow/go-ethereum/common"
+	"github.com/onflow/go-ethereum/eth/tracers"
+
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// NewBlockContext creates a new block context for the given chain ID and height.
+// This is for use in offchain re-execution of transactions.
+// It includes special casing for some historical block heights:
+//   - On Mainnet and Testnet the block hash list was stuck in a loop of 256 block hashes until fixed.
+//     https://github.com/onflow/flow-go/issues/6552
+//   - The coinbase address was different on testnet until https://github.com/onflow/flow-evm-gateway/pull/491.
+func NewBlockContext(
+	chainID flow.ChainID,
+	height uint64,
+	timestamp uint64,
+	getHashByHeight func(uint64) gethCommon.Hash,
+	prevRandao gethCommon.Hash,
+	tracer *tracers.Tracer,
+) (types.BlockContext, error) {
+
+	// coinbase address fix
+	miner := types.CoinbaseAddress
+	if chainID == flow.Testnet && height < coinbaseAddressChangeEVMHeightTestnet {
+		miner = oldCoinbaseAddressTestnet
+	}
+
+	return types.BlockContext{
+		ChainID:                types.EVMChainIDFromFlowChainID(chainID),
+		BlockNumber:            height,
+		BlockTimestamp:         timestamp,
+		DirectCallBaseGasUsage: types.DefaultDirectCallBaseGasUsage,
+		DirectCallGasPrice:     types.DefaultDirectCallGasPrice,
+		GasFeeCollector:        miner,
+		GetHashFunc: func(n uint64) gethCommon.Hash {
+			// For block heights greater than or equal to the current,
+			// return an empty block hash.
+			if n >= height {
+				return gethCommon.Hash{}
+			}
+			// If the given block height, is more than 256 blocks
+			// in the past, return an empty block hash.
+			if height-n > 256 {
+				return gethCommon.Hash{}
+			}
+
+			// For testnet & mainnet, we fetch the block hash from the hard-coded
+			// array of hashes.
+			if chainID == flow.Mainnet && height < blockHashListFixHCUEVMHeightMainnet {
+				return fixedHashes[flow.Mainnet][height%256]
+			} else if chainID == flow.Testnet && height < blockHashListFixHCUEVMHeightTestnet {
+				return fixedHashes[flow.Testnet][height%256]
+			}
+
+			return getHashByHeight(n)
+
+		},
+		Random: prevRandao,
+		Tracer: tracer,
+	}, nil
+}
+
+// TODO: HCU not done yet
+// PR: https://github.com/onflow/flow-go/pull/6734
+const blockHashListFixHCUEVMHeightMainnet = 0
+
+// Testnet52 - Height Coordinated Upgrade 4, Nov 20, 2024
+// Flow Block: 228025500 7eb808b77f02c3e77c36d57dc893ed63adc5ff6113bb0f4b141bb39e44d634e6
+// PR: https://github.com/onflow/flow-go/pull/6734
+const blockHashListFixHCUEVMHeightTestnet = 16848829
+
+// Testnet51 - Height Coordinated Upgrade 1
+// Flow Block: 212562161 1a520608c5457f228405c4c30fc39c8a0af7cf915fb2ede7ec5ccffc2a000f57
+// PR: https://github.com/onflow/flow-go/pull/6380
+// TODO: should this be 1385490?
```

Comment:
@m-Peter @zhangchiqing not sure if this should be the 1st height of the HCU or the last height before the HCU. In the testing we did, this was the last height before the HCU but I dont think that is technically correct.